### PR TITLE
feat(language-core): improve support for range mapping

### LINKED ISSUE #203: Support overlapping mapping ranges
I am continuing my work on Angular support and a lot of stuff is working correctly in version 2.3.0, however there is one problem with transpilation of `?.`, which for historical reasons works differently then TypeScript version of `?.`. Let me describe the problem.

## Problem

WebStorm is using TypeScript server to provide types of expressions. For instance, if we have `foo.bar.fooBar` expression, to get the type of `fooBar` property, WebStorm will check expression type of the whole range `foo.bar.fooBar`. This fails to work with transpiled `?.` though. Currently, the ranges would be incorrectly mapped from the source file to generated file, e.g. `data?.icon` would be mapped to `data)!.icon`. 

Visualization of Angular Template Transpilation 2024-06-17 11-36-46

The problem is that `?.` is expanded to a whole expression, so we need to add more mappings and they will start overlapping:

Visualization of Angular Template Transpilation 2024-06-17 11-38-39

So, if we map source range `data?.icon` we should get `(null as any ? (this.data)!.icon : undefined)` generated range, but if we want to map an error from generated range that `icon` property does not exist on `data`, we should get a different mapping:

Visualization of Angular Template Transpilation 2024-06-17 11-39-51

## Solution

I want to store source mappings in an interval tree (e.g. https://www.npmjs.com/package/@flatten-js/interval-tree). This will allow for a given text range to find all mappings containing a start point and an end point in O(ln(n)) time (where n is number of mappings) and then find among them the best match:
1. If there are mappings containing both start and end point - find the smallest range among them. If start and end match perfectly use the target range as is.
This also solves issues with mapping of e.g. `_t1` to `fooBar` (when lengths of source and generated code are different). Currently you get `_t1` mapped to `foo` instead of `fooBar`.
2. Otherwise find the smallest mapping containing start and the smallest mapping containing end point and use them for mapping start and end separately.

@johnsoncodehk - I can implement the solution and provide a PR with it. Let me know what you think about this!

### PR DESCRIPTION
**Pull request by piotrtomiak**
Fixes #203

### CODE REVIEW

File: packages/language-service/lib/documents.ts [hunk 1]
```diff
@@ -60,13 +60,13 @@ export class SourceMapWithDocuments {
 	}
 
 	public * getSourcePositions(position: vscode.Position, filter: (data: CodeInformation) => boolean = () => true) {
-		for (const mapped of this.findPositions(position, filter, this.embeddedDocument, this.sourceDocument, 'generatedOffsets', 'sourceOffsets')) {
```
Comment (piotrtomiak) [inline packages/language-service/lib/documents.ts:63]: I've cleaned up an API a little bit - 'sourceOffsets' is a duplicated information here

File: packages/language-service/lib/documents.ts [hunk 2]
```diff
@@ -76,10 +76,9 @@ export class SourceMapWithDocuments {
 		filter: (data: CodeInformation) => boolean,
 		fromDoc: TextDocument,
 		toDoc: TextDocument,
-		from: CodeRangeKey,
-		to: CodeRangeKey
+		from: CodeRangeKey
 	) {
-		for (const mapped of this.map.findMatching(fromDoc.offsetAt(position), from, to)) {
+		for (const mapped of this.map.findMatchingOffsets(fromDoc.offsetAt(position), from)) {
```
Comment (piotrtomiak) [inline packages/language-service/lib/documents.ts:81]: Refactored name for consistency with other methods on the `map`

File: packages/language-service/lib/utils/common.ts [hunk 3]
```diff
@@ -11,17 +11,20 @@ export function findOverlapCodeRange(
...
+			end: mappedEnd,
 		}
 	}
-	for (const [mapped, mapping] of map.getGeneratedOffsets(end)) {
-		if (filter(mapping.data)) {
-			mappedEnd = mapped;
-			break;
-		}
+
+	for (const [mapped] of map.getGeneratedOffsets(start, filter)) {
```
Comment (piotrtomiak) [inline packages/language-service/lib/utils/common.ts:21]: `filter` is added to map API, to optimize iterations and disregard mappings early on

File: packages/language-service/tests/findOverlapCodeRange.spec.ts [hunk 4]
```diff
@@ -90,4 +91,34 @@ describe(`Test findOverlapCodeRange()`, () => {
 
 		expect(findOverlapCodeRange(0, 38, map, () => true)).toEqual({ start: 6, end: 33 });
 	});
+
+	it('overlapping ranges', () => {
+		const mappings: Mapping<CodeInformation>[] = [
```
Comment (piotrtomiak) [inline packages/language-service/tests/findOverlapCodeRange.spec.ts:96]: Overlapping ranges actually work, they just need to be in separate mappings

File: packages/source-map/lib/sourceMap.ts [hunk 5]
```diff
@@ -11,76 +11,140 @@ export interface Mapping<Data = unknown> {
...
+		for (const match of this.getStorageBasedOnRange(fromRange).findMatchingMappingIndexes(offset, filter)) {
+			yield [mapOffset(match, offset, fromRange, toRange), match[0]];
+		}
 	}
 
-	* findMatching(offset: number, fromRange: CodeRangeKey, toRange: CodeRangeKey) {
-		const memo = this.getMemoBasedOnRange(fromRange);
-		if (memo.offsets.length === 0) {
-			return;
+	* findMatchingStartEnd(start: number, end: number, fromRange: CodeRangeKey, filter?: (data: Data) => boolean): Generator<[start: number, end: number, Mapping<Data>]> {
```
Comment (piotrtomiak) [inline packages/source-map/lib/sourceMap.ts:45]: I've encapsulated all of the range mapping logic here. The main idea is to not take the first mapping possible, but look for the one which fits the range the best, i.e. is the shortest possible.

File: packages/source-map/tests/sourceMap.spec.ts [hunk 6]
```diff
@@ -149,6 +149,162 @@ describe('sourceMap', () => {
...
+			`{{|data?.icon?.toString()}}`.indexOf('|'),
+			`{{data?.icon?.toString()|}}`.indexOf('|')
+		)].map(mapped => mapped.slice(0, 2))).toEqual([
+			[
+				0,
+				`(null as any ? ((null as any ? ((null as any ? (this.data)!.icon : undefined)!.toString : undefined))!() : undefined)`.length,
+			],
+		]);
+	});
+	test('Angular template with full ranges', () => {
```
Comment (piotrtomiak) [inline packages/source-map/tests/sourceMap.spec.ts:162]: Added a test case without zero-length ranges and mappings with overlapping ranges.

File: packages/source-map/lib/sourceMap.ts [hunk 7]
```diff
@@ -11,76 +11,140 @@ export interface Mapping<Data = unknown> {
...
+	private sourceCodeOffsetsStorage: BinarySearchStorage<Data> | undefined;
+	private generatedCodeOffsetsStorage: BinarySearchStorage<Data> | undefined;
 
-	constructor(public readonly mappings: Mapping<Data>[]) { }
+	constructor(public readonly mappings: Mapping<Data>[]) {
+	}
 
-	getSourceStartEnd(generatedStart: number, generatedEnd: number) {
-		return this.findMatchingStartEnd(generatedStart, generatedEnd, 'generatedOffsets', 'sourceOffsets');
+	getSourceStartEnd(generatedStart: number, generatedEnd: number, filter?: (data: Data) => boolean): Generator<[start: number, end: number, Mapping<Data>]> {
```
Comment (piotrtomiak) [inline packages/source-map/lib/sourceMap.ts:22]: Besides refactoring API a bit the main change is in `findMatchingStartEnd`

File: packages/source-map/lib/sourceMap.ts [hunk 8]
```diff
@@ -11,76 +11,140 @@ export interface Mapping<Data = unknown> {
...
+	private createStorage(key: CodeRangeKey): BinarySearchStorage<Data> {
+		if (!this.mappings.every(mapping => areRangesSortedAndNonOverlapping(mapping[key], getLengths(mapping, key)))) {
+			throw new Error("Ranges within one mapping should be sorted and non-overlapping.")
+		}
+		return new BinarySearchStorage<Data>(this.mappings, key)
+	}
+
+}
+
+class BinarySearchStorage<Data> {
```
Comment (piotrtomiak) [inline packages/source-map/lib/sourceMap.ts:140]: I've originally had the interval storage here, which turned out to be inefficient and not really needed anyway, so I kept the logic in the separate class.

File: packages/source-map/lib/sourceMap.ts [hunk 9]
```diff
@@ -99,11 +163,78 @@ export class SourceMap<Data = unknown> {
...
+				let low = 0;
+				let high = fromOffsets.length - 1;
+
+				while (low <= high) {
+					const mid = Math.floor((low + high) / 2);
+					const fromOffset = fromOffsets[mid];
+					const fromLength = fromLengths[mid];
+
+					if (offset >= fromOffset && offset <= fromOffset + fromLength) {
+						yield [mapping, mid]
```
Comment (piotrtomiak) [inline packages/source-map/lib/sourceMap.ts:201]: The main difference is that we don't return offset, but the mapping with range index. This allows us to find ranges which contain both start an end and prefer them.

File: packages/source-map/lib/translateOffset.ts [hunk 10]
```diff
@@ -24,3 +23,14 @@ export function translateOffset(start: number, fromOffsets: number[], toOffsets:
 		}
 	}
 }
+
+export function areRangesSortedAndNonOverlapping(offsets: number[], lenghts: number[]): boolean {
```
Comment (piotrtomiak) [inline packages/source-map/lib/translateOffset.ts:27]: A little bit stricter check - no overlapping allowed within a particular mapping

File: packages/typescript/lib/node/decorateLanguageService.ts [hunk 11]
```diff
@@ -136,10 +135,9 @@ export function decorateLanguageService(
 			return [];
 		}
 		if (serviceScript) {
-			const generateStart = toGeneratedOffset(language, serviceScript, sourceScript, start, isFormattingEnabled);
-			const generateEnd = toGeneratedOffset(language, serviceScript, sourceScript, end, isFormattingEnabled);
-			if (generateStart !== undefined && generateEnd !== undefined) {
-				const edits = getFormattingEditsForRange(targetScript.id, generateStart, generateEnd, options);
+			const generatedRange = toGeneratedRange(language, serviceScript, sourceScript, { pos: start, end }, isCodeActionsEnabled);
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/decorateLanguageService.ts:138]: Use range

File: packages/typescript/lib/node/decorateLanguageService.ts [hunk 12]
```diff
@@ -421,8 +419,8 @@ export function decorateLanguageService(
 				}
 			}
 			else {
-				for (const [generatedStart, generatedEnd] of toGeneratedRanges(language, serviceScript, sourceScript, positionOrRange.pos, positionOrRange.end, isCodeActionsEnabled)) {
-					return getApplicableRefactors(targetScript.id, { pos: generatedStart, end: generatedEnd }, preferences, triggerReason, kind, includeInteractiveActions);
+				for (const generatedRange of toGeneratedRanges(language, serviceScript, sourceScript, positionOrRange, isCodeActionsEnabled)) {
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/decorateLanguageService.ts:422]: I've changed API of toGenerateRanges to return a range object

**Comment by piotrtomiak** [comment]
Hi! I've added some refinements to the startEnd matching logic. It is more readable now and also it supports the fallback mode out-of-the-box.

**Comment by johnsoncodehk** [comment]
Since overlapping mapping handling is Angular-specific behavior, I think it's best to keep this implementation in WebStorm-Angular and keep the default implementation simple.

#207 Added a method to make mapping logic customizable, can you check if it meets your needs?

**Comment by piotrtomiak** [comment]
@johnsoncodehk - sounds good! I've copy/paste my implementation of source map to angular plugin and it works fine! Thanks for the changes!

**Pull request closed without merging**